File: solutions/lua/affine-cipher/1/affine-cipher.lua
local function chunk(str)
	local res = ""
	for i = 1, #str do
		res = res .. str:sub(i, i)
		if i % 5 == 0 and i ~= #str then
			res = res .. " "
		end
	end
	return res
end

local alphabet = "abcdefghijklmnopqrstuvwxyz"

local letters = {}
for i = 1, #alphabet do
	local c = alphabet:sub(i, i)
	local C = string.upper(c)
	letters[c] = i-1
	letters[i-1] = c
	letters[C] = i-1
end

local m = #alphabet

local function all(xs, f)
	for _, x in ipairs(xs) do
		if not f(x) then
			return false
		end
	end
	return true
end

local function tot(n)
	local primes = {}
	local res = 1
	for i = 2, n do
		if all(primes, function(p) return i % p ~= 0 end) and n % i == 0 then
			table.insert(primes, i)
			res = res * (1 - (1 / i))
		end
	end
	return res * n
end

local function mmi(a, m)
	return math.floor(a ^ (tot(m) - 1))
end

local function coprime(x, y)
	for i = 2, math.min(x, y) do
		if x % i == 0 and y % i == 0 then
			return false
		end
	end
	return true
end

local function filter(str)
	local res = ""
	for c in str:gmatch("[^%s,.]") do
		res = res .. c
	end
	return res
end

local function encode(phrase, key)
	if not coprime(key.a, m) then
		error("a and m must be coprime.")
	end
	return chunk(filter(phrase):gsub('.', function(x)
		if string.match(x, "%d") then
			return x
		else
			return letters[(key.a * letters[x] + key.b) % m]
		end
	end))
end

local function decode(phrase, key)
	if not coprime(key.a, m) then
		error("a and m must be coprime.")
	end
	return filter(phrase):gsub('.', function(y)
		if string.match(y, "%d") then
			return y
		else
			return letters[(mmi(key.a, m) * (letters[y] - key.b)) % m]
		end
	end)
end

return { encode = encode, decode = decode }

File: solutions/cpp/affine-cipher/1/affine_cipher.cpp
#include "affine_cipher.h"

#include <algorithm>
#include <cassert>
#include <cctype>
#include <cmath>
#include <locale>
#include <stdexcept>
#include <string_view>

namespace affine_cipher {

constexpr const std::string_view alphabet { "abcdefghijklmnopqrstuvwxyz" };
constexpr const auto m = alphabet.size();

static const std::locale loc {"C"};

static const auto filter = [](auto c){
	return c == ',' or c == '.' or std::isspace(c, loc);
};

std::string chunk(std::string str) {
	std::string res {};
	for (auto i = 1; i <= (int)str.size(); i++) {
		res += str[i-1];
		if (i % 5 == 0 and i != (int)str.size())
			res += " ";
	}
	return res;
}

int mmi(int a, int m) {
	for (auto x = 1; x <= m; x++)
		if ((a*x) % m == 1)
			return x;
	throw std::domain_error("no mmi found");
}

bool coprime(int x, int y) {
	for (auto i = 2; i <= std::min(x, y); i++)
		if ((x % i) == 0 and (y % i) == 0)
			return false;
	return true;
}

int euc(int a, int b) {
	return ((a % b) + b) % b;
}

std::string encode(std::string phrase, int a, int b) {
	if (not coprime(a, m))
		throw std::invalid_argument("a and m must be coprime");
	phrase.erase(std::remove_if(phrase.begin(), phrase.end(), filter), phrase.end());
	std::transform(phrase.begin(), phrase.end(), phrase.begin(), [&](auto c) {
		return std::isalpha(c, loc)
			? (a * (std::tolower(c, loc) - 'a') + b) % m + 'a'
			: c;
	});
	return chunk(phrase);
}

std::string decode(std::string phrase, int a, int b) {
	if (not coprime(a, m))
		throw std::invalid_argument("a and m must be coprime");
	phrase.erase(std::remove_if(phrase.begin(), phrase.end(), filter), phrase.end());
	std::transform(phrase.begin(), phrase.end(), phrase.begin(), [&](auto c) {
		return std::isalpha(c, loc)
			? euc(mmi(a, m) * ((c - 'a') - b), m) + 'a'
			: c;
	});
	return phrase;
}

}  // namespace affine_cipher
